[debug-info] Only hoist the first dbg inst associated with an async debug variable.

gottesmm opened the pull request:
Otherwise, we may hoist a llvm.dbg.addr over another llvm.dbg.addr or
llvm.dbg.value, changing observable debugging state.

As an additional benefit, this finally lets me add straight line end to end lldb
tests for async var and lets. This was the last thing stopping me from doing that.

gottesmm committed:
[debug-info] Only hoist the first dbg inst associated with an async debug variable.

Otherwise, we may hoist a llvm.dbg.addr over another llvm.dbg.addr or
llvm.dbg.value, changing observable debugging state.

As an additional benefit, this finally lets me add straight line end to end lldb
tests for async var and lets. This was the last thing stopping me from doing that.

gottesmm commented:
This is a piece of code that AdrianP and I already looked at together. I just split it off and added some more tests.

gottesmm commented:
@swift-ci test

adrian-prantl commented on lldb/test/API/lang/swift/variables/move_function_async/TestSwiftMoveFunctionAsync.py:
```diff
@@ -0,0 +1,171 @@
+        self.do_setup_breakpoints()
+
+        self.process = self.target.LaunchSimple(None, None, os.getcwd())
+        threads = lldbutil.get_threads_stopped_at_breakpoint(
```
Can you replace this boilerplate with a call to 
```
        target, process, thread, bkpt = lldbutil.run_to_source_breakpoint(
            self, 'Set breakpoint', lldb.SBFileSpec('main.swift'))
```
? It doesn't look like the numbering of the breakpoints is actually used, so why not simplify it.

adrian-prantl commented on lldb/test/API/lang/swift/variables/move_function_async/TestSwiftMoveFunctionAsync.py:
```diff
@@ -0,0 +1,171 @@
+
+    def get_var(self, name):
+        frame = self.thread.frames[0]
+        self.assertTrue(frame.IsValid(), "Couldn't get a frame.")
```
run_to_source_breakpoint will check this for you

gottesmm replied:
@adrian-prantl do you want me to run to source breakpoint instead of continuing?

adrian-prantl replied:
If you use run_to_source_breakpoint it will set all breakpoints matching the regex and stop at the first one. If it doesn't raise an exception it's safe assume that frame is valid.

gottesmm replied:
I actually can't do that in this PR for this test. It is b/c there is a stepping issue where we step backwards due to an instruction having a SILLocation that is out of order. The information around what values are alive is different there and so the test fails. If possible, I would like to get this in to test out the actual functionality and then loop back around and figure out the stepping issue.

gottesmm replied:
Sorry, let me be a little clearer. To work around this issue I was disabling that earlier break point and at least my method of disabling from earlier failed. I guess the real question is... do you have any thoughts on how to work around that using run_to_source_breakpoint?

adrian-prantl commented on lldb/test/API/lang/swift/variables/move_function_async/TestSwiftMoveFunctionAsync.py:
```diff
@@ -0,0 +1,171 @@
+        self.assertEqual(varK.unsigned, 0, "varK initialized too early?!")
+
+        # Go to break point 2.1. k should be valid.
+        self.runCmd('continue')
```
It's preferred to use the API for this: `process.Continue()`

gottesmm replied:
Ah. I didn't know about that! Thanks!

gottesmm replied:
I am going to fix that in all of the tests here so I don't have to do another PR.

adrian-prantl commented on llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp:
```diff
@@ -2197,12 +2197,19 @@ bool VarLocBasedLDV::ExtendRanges(MachineFunction &MF,
+      if (SeenDebugVars.insert(DebugVariable(MI.getDebugVariable(),
+                                             MI.getDebugExpression(),
+                                             MI.getDebugLoc()->getInlinedAt())).second  &&
+          isSwiftAsyncContext(MI)) {
```
this lgtm

gottesmm commented:
@swift-ci test

gottesmm committed:
[lldb] Gardening: When testing use self.process.Continue() instead of handrolling continue via a run cmd.

gottesmm commented:
apple/swift#42169
@swift-ci test

gottesmm commented:
"m" and "k" in a specific test flipped order. So I had to edit a test a little bit.

gottesmm committed:
[lldb] Simplify logic in a test by using lldbutil.run_to_source_breakpoint.

gottesmm commented:
https://github.com/apple/swift/pull/42169
@swift-ci test

gottesmm commented on lldb/test/API/lang/swift/variables/move_function_async/TestSwiftMoveFunctionAsync.py:
```diff
@@ -0,0 +1,171 @@
+        #
+        # TODO: Investigate why this is happening!
+        self.runCmd('# Skipping bad loc by disabling earlier break point 6')
+        self.runCmd('break dis 6')
```
@adrian-prantl this is the place where we hit the bad loc.

gottesmm committed:
[lldb] When matching entry values for async swift function, fallback to the current function from parent function when evaluating subexpressions.

This occurs when the parent function does not have debug info (*). In swift async
functions this is ok since we only need to be able to restore the ABI defined
context register for the async context (r14)... so it is not necessary to match
with a callsite parameter.

(*) E.x.: On Linux async funclets called from a libswift_Concurrency.so that had its
debug info stripped.

gottesmm commented:
So I tracked down the linux failure with AdrianP offline. The issue is a separate bug in lldb where we fail to be able to lookup entry values if our parent func doesn't have debug info. For our purposes here for entry values this isn't true due to our ability to rely on ABI. So in that case, we can just rely on the current function having debug info. I together with AdrianP put together a patch that does this. I am going to commit it with this PR since otherwise the linux test fails.

gottesmm commented:
https://github.com/apple/swift/pull/42169
@swift-ci test

gottesmm commented:
https://github.com/apple/swift/pull/42169
@swift-ci test
